pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+		return true
+	}
+	return (s[i].CreationTimestamp == s[j].CreationTimestamp && s[i].UID < s[i].UID)
+}
+
+// processInitialSync processes the items queued by informer's initial sync.
+// A mutex is shared between this method and enqueueService to ensure
+// that queue additions are blocked while the sync is being processed.
+// Returns a boolean indication of whether processing should continue.
+func (ic *IngressIPController) processInitialSync() bool {
+	ic.mu.Lock()
+	defer ic.mu.Unlock()
+
+	glog.V(4).Infof("Processing initial sync")
+
+	// Drain the initial queue.  Len() should be safe because
+	// enqueueService requires the same lock held by this method.
+	for ic.queue.Len() > 0 {
+		if !ic.work() {
+			return false
+		}
+	}
+
+	// Add unallocated services back to the queue in consistent order.
+	sort.Sort(serviceAge(ic.pendingAllocations))
+	for _, service := range ic.pendingAllocations {
+		key, err := controller.KeyFunc(service)
+		if err != nil {
+			// This error should have been caught by enqueueService
+			glog.Errorf("Couldn't get key for service %+v: %v", service, err)
+			continue
+		}
+		glog.V(4).Infof("Adding service to allocation queue: %v ", key)
+		change := &serviceChange{key: key}
+		ic.queue.Add(change)
+	}
+
+	ic.pendingAllocations = nil
+	ic.syncProcessed = true
+
+	glog.V(4).Infof("Completed processing initial sync")
+
+	return true
+}
+
+// worker runs a worker thread that dequeues items, processes them,
+// and marks them done.
+func (ic *IngressIPController) worker() {
+	for {
+		if !ic.work() {
+			return
+		}
+	}
+}
+
+// work returns true if the worker thread should continue
+func (ic *IngressIPController) work() bool {
+	change, quit := ic.queue.Get()
+	if quit {
+		return false
+	}
+	defer ic.queue.Done(change)
+
+	if err := ic.changeHandler(change.(*serviceChange)); err == nil {
+		// No further processing required
+		ic.queue.Forget(change)
+	} else {
+		// Failed but can be retried
+		utilruntime.HandleError(fmt.Errorf("error syncing service, it will be retried: %v", err))
+		ic.queue.AddRateLimited(change)
+	}
+
+	return true
+}
+
+// processChange responds to a service change
+// (addition/update/deletion) by synchronizing the local and persisted
+// ingress ip allocation state of the service.
+func (ic *IngressIPController) processChange(change *serviceChange) error {
+	var service *kapi.Service
+	if change.key != "" {
+		obj, exists, err := ic.cache.GetByKey(change.key)
+		if err != nil {
+			glog.V(4).Infof("Unable to retrieve service %v from store: %v", change.key, err)
+		} else if !exists {
+			glog.V(4).Infof("Service has been deleted %v", change.key)
+		} else {
+			service = obj.(*kapi.Service)
+		}
+	}
+
+	oldIP := ""
+	if change.oldService != nil && change.oldService.Spec.Type == kapi.ServiceTypeLoadBalancer {
+		ingress := change.oldService.Status.LoadBalancer.Ingress
+		if len(ingress) > 0 {
+			oldIP = ingress[0].IP
+		}
+	}
+
+	if service != nil {
```

Comment:
Anytime we have a big nested if / else, it should usually be a `switch`.  In this case, 

```
switch {
case service == nil && len(oldIP) > 0:
  // ...
case service == nil:
  // ...
case service.Spec.Type != ServiceTypeLoadBalancer:
...
}
```

Which ends up forcing someone to walk through the cases more clearly.